src/lib/transition_router/initial_validator.ml

```diff
@@ -48,17 +48,53 @@ module Make (Inputs : Transition_frontier.Inputs_intf) = struct
              , [("slot_diff", `String (Int64.to_string slot_diff))] ))
 
   module Duplicate_proposal_detector = struct
-    module T = struct
-      (* int is a slot *)
-      type t = Public_key.Compressed.Stable.V1.t * int
-      [@@deriving bin_io, sexp, hash, compare]
+    (* maintain a map from proposer, epoch, slot to state hashes *)
+
+    module Proposals = struct
+      module T = struct
+        (* order of fields significant, compare by epoch, then slot, then proposer *)
+        type t =
+          {epoch: int; slot: int; proposer: Public_key.Compressed.Stable.V1.t}
+        [@@deriving sexp, compare]
+      end
+
+      include T
+      include Comparable.Make (T)
     end
 
-    module Key = Hashable.Make_binable (T)
+    type t =
+      {mutable table: State_hash.t Proposals.Map.t; mutable latest_epoch: int}
 
-    type t = {table: State_hash.t Key.Table.t; mutable latest_epoch: int}
+    (* every gc_interval proposals seen, discard proposals more than gc_width ago *)
+    let table_gc =
+      let open Consensus in
+      let delay = Data.Consensus_state.network_delay Configuration.t in
+      let gc_width = delay * 2 in
+      let gc_interval = gc_width * 2 in
+      let count = ref 0 in
+      (* create dummy proposal to split map on *)
+      let make_splitting_proposal (proposal : Proposals.t) : Proposals.t =
+        let proposer = Public_key.Compressed.empty in
+        if proposal.slot >= gc_width then
+          (* within current epoch *)
+          {epoch= proposal.epoch; slot= proposal.slot - gc_width; proposer}
+        else if Int.equal proposal.epoch 0 then
+          (* back to the beginning *)
+          {epoch= 0; slot= 0; proposer}
+        else
+          (* in previous epoch *)
+          { epoch= proposal.epoch - 1
+          ; slot= epoch_size - gc_width + proposal.slot
+          ; proposer }
+      in
+      fun t proposal ->
```

Comment:
Can we have this as explicit arguments of when you are declaring the function? I was a bit confused when you did this.

So I prefer you write something like:

```
let table_gc t proposal = ...
```